core/node/crypto/client.go

```diff
@@ -2,140 +2,404 @@ package crypto
-	ctx, span := ic.tracer.Start(ctx, "eth_getHeaderByNumber")
-	defer span.End()
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_getHeaderByNumber")
+		defer span.End()
+	}
 
 	return ic.Client.HeaderByNumber(ctx, number)
 }
 
 func (ic *otelEthClient) BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_getBlockByNumber")
-	defer span.End()
-
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_getBlockByNumber")
+		defer span.End()
+	}
 	return ic.Client.BlockByNumber(ctx, number)
 }
 
-func (ic *otelEthClient) CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_callContract")
-	defer span.End()
-
-	if len(msg.Data) >= 4 {
-		span.SetAttributes(attribute.String("func_selector", hex.EncodeToString(msg.Data[:4])))
+// extractCallErrorStatus extracts the revert reason from an error if it is a contract error
+// with a revert reason. Otherwise, it will return "revert" for contract data errors, or
+// "fail" for other types of errors.
+func extractCallErrorStatus(err error) string {
+	if de, ok := err.(rpc.DataError); ok {
+		hexStr := de.ErrorData().(string)
+		hexStr = strings.TrimPrefix(hexStr, "0x")
+		revert, e := hex.DecodeString(hexStr)
+		if e == nil {
+			reason, e := abi.UnpackRevert(revert)
+			if e == nil {
+				return reason
+			}
+		}
+		return "revert"
 	}
+	return "fail"
+}
 
-	return ic.Client.CallContract(ctx, msg, blockNumber)
+func getMethodName(data *[]byte) string {
+	var methodName string
+	if len(*data) > 4 {
+		selector := binary.BigEndian.Uint32(*data)
+		var defined bool
+		methodName, defined = GetSelectorMethodName(selector)
+		if !defined {
+			return hex.EncodeToString((*data)[:4])
+		}
+	}
+	return methodName
 }
 
-func (ic *otelEthClient) CallContractAtHash(ctx context.Context, msg ethereum.CallMsg, blockHash common.Hash) ([]byte, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_callContractAtHash")
-	defer span.End()
+func (ic *otelEthClient) makeEthCallWithTraceAndMetrics(
+	ctx context.Context,
+	msg ethereum.CallMsg,
+	call func() ([]byte, error),
+) ([]byte, error) {
+	var methodName string
+
+	if ic.tracer != nil {
+		var span trace.Span
+		_, span = ic.tracer.Start(ctx, "eth_call")
+		defer span.End()
+
+		methodName = getMethodName(&msg.Data)
+		span.SetAttributes(attribute.String("method_name", methodName))
+	}
 
-	if len(msg.Data) >= 4 {
-		span.SetAttributes(attribute.String("func_selector", hex.EncodeToString(msg.Data[:4])))
+	data, err := call()
+
+	if ic.ethCalls != nil {
+		status := "ok"
+		if err != nil {
+			status = extractCallErrorStatus(err)
+		}
+
+		if methodName == "" {
+			methodName = getMethodName(&msg.Data)
+		}
+		ic.ethCalls.With(
+			prometheus.Labels{
+				"chain_id":    ic.chainId,
+				"method_name": methodName,
```

Comment:
The txpool labels transaction metrics with `func_selector`. If the read operations move to this new convention with method name the txpool metrics should also move to this new convention. And as a result the txpool dashboard in dd needs to be updated as well.